src/entity/aggregator/new_aggregator.rs

```diff
@@ -0,0 +1,69 @@
+use crate::{
+    entity::{url_safe_base64, Account},
+    handler::Error,
+};
+use sea_orm::Set;
+use serde::{Deserialize, Serialize};
+use std::str::FromStr;
+use time::OffsetDateTime;
+use uuid::Uuid;
+use validator::{Validate, ValidationError, ValidationErrors};
+
+use super::{ActiveModel, Role};
+
+#[derive(Deserialize, Serialize, Validate, Debug, Clone, Default)]
+pub struct NewAggregator {
+    #[validate(required, custom = "validate_role")]
+    pub role: Option<String>,
+    #[validate(required, length(min = 1))]
+    pub name: Option<String>,
+    #[validate(url)]
+    pub api_url: Option<String>,
+    #[validate(required, url)]
+    pub dap_url: Option<String>,
+    #[validate(custom = "url_safe_base64")]
+    pub bearer_token: Option<String>,
+}
+
+fn validate_role(role: &str) -> Result<(), ValidationError> {
+    Role::from_str(role)
+        .map_err(|_| ValidationError::new("role"))
+        .map(|_| ())
+}
+
+impl NewAggregator {
+    pub fn validate(&self) -> Result<(), ValidationErrors> {
+        let errors = Validate::validate(self);
+        match (&self.dap_url, &self.bearer_token) {
+            (Some(_), Some(_)) | (None, None) => errors,
+            (Some(_), None) => {
+                let mut err = errors.err().unwrap_or_default();
+                err.add("bearer_token", ValidationError::new("required"));
+                Err(err)
+            }
+
+            (None, Some(_)) => {
+                let mut err = errors.err().unwrap_or_default();
+                err.add("dap_url", ValidationError::new("required"));
+                Err(err)
+            }
+        }
+    }
+
+    pub fn build(self, account: Option<&Account>) -> Result<ActiveModel, Error> {
+        self.validate()?;
+
+        Ok(ActiveModel {
+            role: Set(self.role.unwrap().parse().unwrap()),
+            name: Set(self.name.unwrap()),
+            api_url: Set(self.api_url.map(|u| u.parse()).transpose()?),
+            dap_url: Set(self.dap_url.unwrap().parse()?),
```

Comment:
I'm a little concerned with all the `.unwrap()` here. Since this method returns an error anyway, could we just propagate errors up? Or maybe we need different types for the JSON message on the wire (where some fields are `Option<whatever>`) and the representation of the message we work with after parsing (where fields are required)?

Reply:
This code is in fact transitioning from the optional json message on the wire to the representation after validation. L54 calls self.validate, which means that none of these unwraps are reachable if the record is missing any of the required fields. As such, the panic arm of the unwrap is unreachable.

Reply:
We have similar logic for `NewTask` and `NewAccount`

Reply:
Fair enough. In that case, it'd be nice to put in `// unwrap safety:` comments explaining why these `unwrap()`s are harmless, which we try to do in Janus and our other Rust projects.

Reply:
addressed in e4e640e (and further convinced myself that a critical library is missing from the rust ecosystem)